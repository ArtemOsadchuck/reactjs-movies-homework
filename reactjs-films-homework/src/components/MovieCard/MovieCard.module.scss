$background-card-color: rgb(0, 0, 0);
$background-img-color: rgb(231, 231, 231);
$background-rating-color: rgb(41, 186, 41);
$play-icon-size: 44px;
$wite-color: rgb(255, 255, 255);

.movieCardLink {
  align-items: center;
  display: flex;
  fill: rgb(201, 30, 30);
  flex-direction: column;
  height: max-content;
  margin: 0;
  min-height: 24.5rem;
  overflow: hidden;
  position: relative;
  width: 13rem;

  span {
    font-size: 0.6rem;
    text-align: start;
  }

  .voteAverage {
    font-size: 0.6rem;
    opacity: 1;
  }

  .imgCardWrapper {
    align-items: center;
    background-color: $background-img-color;
    border-radius: 5px;
    display: flex;
    height: 17rem;
    justify-content: center;
    margin: 5% auto;
    transition: 0.5s ease-in-out;
    width: 98%;
  }

  h4 {
    font-weight: 400;
    letter-spacing: 1px;
    padding-left: 5%;
    text-align: start;
  }

  .ratingMovie {
    align-items: center;
    background-color: $background-rating-color;
    border-radius: 3px;
    color: $wite-color;
    display: flex;
    height: 1rem;
    justify-content: center;
    left: 0.5rem;
    overflow: hidden;
    position: absolute;
    top: 0.9rem;
    width: 1.8rem;
    z-index: 11;
  }

  .playIconInner {
    border-color: transparent transparent transparent $wite-color;
    border-style: solid;
    border-width: 8px 0 8px 12px;
    display: inline-block;
    height: 0;
    margin-left: 5%;
    width: 0;
  }

  .playIcon {
    align-items: center;
    background-color: $background-rating-color;
    border-radius: 50%;
    display: flex;
    height: $play-icon-size;
    justify-content: center;
    left: 50%;
    margin-bottom: 2px;
    opacity: 0;
    position: absolute;
    top: 11rem;
    transform: translate(-49%, -120%);
    transition: 0.5s ease-in-out;
    width: $play-icon-size;
    z-index: 11;
  }

  .movieCardWrapper {
    align-items: center;
    border-radius: 2px;
    color: $background-img-color;
    letter-spacing: 1px;
    min-height: 20rem;
    overflow: hidden;
    padding: 0 4px;
    position: relative;
    text-align: start;
    width: 12rem;
    z-index: 10;
  }

  .genre {
    margin: 0;
    padding-left: 5%;
    text-align: start;
  }

  &:hover {
    .imgCardWrapper {
      opacity: 0.3;
    }

    .playIcon {
      opacity: 1;
    }

    h4 {
      color: $wite-color;
    }

    span {
      color: $background-img-color;
    }
  }
}
