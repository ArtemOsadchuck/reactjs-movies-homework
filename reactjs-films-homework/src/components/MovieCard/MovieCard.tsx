/* eslint-disable array-callback-return */
import React, { useState, useMemo } from 'react';
import img from './img/Movie-card.png';
import styles from './MovieCard.module.scss';
import { useAppSelector, useAppDispatch } from '../../hooks/hooks';
import { setMovieID } from '../../store/rootStore/movieDetailsPageStore/movieDetailsPageSlice';
import GenresList from './GenresList';

export interface ICard {
  props: IMovieCard;
}

export interface IMovieCard {
  id: number;
  title: string;
  vote_average: number;
  poster_path: string;
  genre_ids?: number[];
  key?: number;
}

export interface IGenre {
  genres?: {
    id: number;
    name: string;
  }[];
}

const MovieCard: React.FC<ICard> = ({
  props: { id, title, vote_average, poster_path, genre_ids },
}) => {
  const imgWidth = '52px';
  const movieTitle = 'Movie Title';
  const urlImg = `https://image.tmdb.org/t/p/w500/${poster_path}`;
  const [genreName, setGenreName] = useState<string[]>([]);
  const dispatch = useAppDispatch();
  const appFetchMovieGenre = useAppSelector((state) => state.mainReducer.genre);

  useMemo(() => {
    const arrToSort: string[] = [];
    if (appFetchMovieGenre!.genres) {
      appFetchMovieGenre!.genres.map((genre): void => {
        genre_ids!.map((el): void => {
          if (genre!.id === el) {
            setGenreName((prev) => {
              arrToSort.push(genre.name);
              const set = new Set(arrToSort);
              const arr = Array.from(set);
              prev = arr;
              return prev;
            });
          }
        });
      });
    }
  }, [genre_ids, appFetchMovieGenre]);

  const voteAverage = useMemo(
    () => Math.ceil(vote_average * 10) / 10,
    [vote_average]
  );

  const goToMovie = (movieID: string) => {
    dispatch(setMovieID(movieID));
  };

  return (
    <div
      id={`${id}`}
      className={styles.movieCardLink}
      onClick={() => goToMovie(`${id}`)}
    >
      <div className={styles.ratingMovie}>
        <p className={styles.voteAverage}>{voteAverage}</p>
      </div>
      <div className={styles.playIcon}>
        <div className={styles.playIconInner}></div>
      </div>
      <div className={styles.movieCardWrapper}>
        <div className={styles.imgCardWrapper}>
          {poster_path ? (
            <img src={urlImg} height="100%" alt={movieTitle} />
          ) : (
            <img src={img} width={imgWidth} alt={movieTitle} />
          )}
        </div>
        <h4>{title}</h4>
        {genreName && <GenresList genreName={genreName} />}
      </div>
    </div>
  );
};

export default MovieCard;
