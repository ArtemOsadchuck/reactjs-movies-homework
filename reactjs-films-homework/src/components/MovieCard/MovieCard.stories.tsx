import React from 'react';

import { Story } from '@storybook/react';
import MovieCard from './MovieCard';

export default {
  title: 'MovieCard',
  component: MovieCard,
};

const MovieCardMocks = {
  adult: false,
  backdrop_path: '/gNBCvtYyGPbjPCT1k3MvJuNuXR6.jpg',
  genre_ids: [35, 18, 10749],
  id: 19404,
  original_language: 'hi',
  original_title: 'दिलवाले दुल्हनिया ले जायेंगे',
  overview:
    'Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.',
  popularity: 16.457,
  poster_path: '/2CAL2433ZeIihfX1Hb2139CX0pW.jpg',
  release_date: '1995-10-20',
  title: 'Dilwale Dulhania Le Jayenge',
  video: false,
  vote_average: 8.7,
  vote_count: 2932,
};
const MovieCardMocksWithoutImg = {
  adult: false,
  backdrop_path: '/gNBCvtYyGPbjPCuNuXR6.jpg',
  genre_ids: [35, 18, 10749],
  id: 19404,
  original_language: 'hi',
  original_title: 'दिलवाले दुल्हनिया ले जायेंगे',
  overview:
    'Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.',
  popularity: 16.457,
  poster_path: '',
  release_date: '1995-10-20',
  title: 'Dilwale Dulhania Le Jayenge',
  video: false,
  vote_average: 8.7,
  vote_count: 2932,
};

const TemplateUrl: Story<typeof MovieCard> = () => (
  <MovieCard props={MovieCardMocks} />
);
const Template: Story<typeof MovieCard> = () => (
  <MovieCard props={MovieCardMocksWithoutImg} />
);

export const MovieCardImg = TemplateUrl.bind({});
export const MovieCardMocksWithoutImgUrl = Template.bind({});
