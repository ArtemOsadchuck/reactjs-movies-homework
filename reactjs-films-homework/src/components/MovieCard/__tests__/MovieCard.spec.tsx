import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import MovieCard from '../MovieCard';

describe('MovieCard', () => {
  const mockProps = {
    id: 724089,
    original_title: "Gabriel's Inferno Part II",
    overview:
      "Professor Gabriel Emerson finally learns the truth about Julia Mitchell's identity, but his realization comes a moment too late. Julia is done waiting for the well-respected Dante specialist to remember her and wants nothing more to do with him. Can Gabriel win back her heart before she finds love in another's arms?",
    popularity: 7.824,
    poster_path: '/x5o8cLZfEXMoZczTYWLrUo1P7UJ.jpg',
    release_date: '2020-07-31',
    title: "Gabriel's Inferno Part II",
    vote_average: 8.7,
  };

  it('MovieCard snapshot', () => {
    const { asFragment } = render(<MovieCard props={mockProps} />);
    expect(asFragment).toMatchSnapshot();
  });

  it('MovieCard Must include rating', () => {
    render(<MovieCard props={mockProps} />);
    expect(screen.getByText(/[0-9]/i)).toBeInTheDocument();
  });
  it('LanguageToggler Must include text RU or EN', () => {
    render(<MovieCard props={mockProps} />);
    expect(screen.getByRole('link')).toBeInTheDocument();
  });
});
