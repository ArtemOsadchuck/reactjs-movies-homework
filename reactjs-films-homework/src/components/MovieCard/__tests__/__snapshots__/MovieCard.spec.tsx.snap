// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard MovieCard snapshot 1`] = `[Function]`;
